airbyte-integrations/connectors/source-intercom/source_intercom/source.py

```diff
@@ -0,0 +1,364 @@
+from urllib.parse import parse_qsl, urlparse
+
+import requests
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator, TokenAuthenticator
+
+
+class IntercomStream(HttpStream, ABC):
+    url_base = "https://api.intercom.io/"
+
+    # https://developers.intercom.com/intercom-api-reference/reference#rate-limiting
+    queries_per_hour = 1000  # 1000 queries per hour == 1 req in 3,6 secs
+
+    primary_key = "id"
+    data_fields = ["data"]
+
+    def __init__(
+        self,
+        authenticator: HttpAuthenticator,
+        start_date: str = None,
+        **kwargs,
+    ):
+        self.start_date = start_date
+
+        super().__init__(authenticator=authenticator)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        Abstract method of HttpStream - should be overwritten.
+        Returning None means there are no more pages to read in response.
+        """
+
+        next_page = response.links.get("next", None)
+
+        if next_page:
+            return dict(parse_qsl(urlparse(next_page.get("url")).query))
+        else:
+            return None
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {"Accept": "application/json"}
+
+    def _send_request(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+        try:
+            return super()._send_request(request, request_kwargs)
+        except requests.exceptions.HTTPError as e:
+            error_message = e.response.text
+            if error_message:
+                self.logger.error(f"Stream {self.name}: {e.response.status_code} " f"{e.response.reason} - {error_message}")
+            raise e
+
+    def get_data(self, response: requests.Response) -> List:
+        data = response.json()
+
+        for data_field in self.data_fields:
+            if data and isinstance(data, dict):
+                data = data.get(data_field, [])
+
+        if isinstance(data, list):
+            data = data
+        elif isinstance(data, dict):
+            data = [data]
+
+        return data
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+
+        data = self.get_data(response)
+
+        for record in data:
+            yield record
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        wait_time = 3600 / self.queries_per_hour  # wait for 3,6 seconds according to API limit
```

Comment:
FYI this does not backoff unless it received a 429 HTTP status code from the server -- is this how we want it to work? 

We could potentially put this as a `time.sleep` inside `parse_response` at the end, since that method is called  once per HTTP request

Reply:
We should also just handle this in #2742

Reply:
Done